horovod/mxnet/__init__.py

```diff
@@ -169,15 +179,22 @@ def _allreduce_grads(self):
                     grads, names = zip(*entries)
                     grouped_allreduce_(tensors=grads, average=False, name="{}:{}".format(names[0], names[-1]), priority=-i,
                                        prescale_factor=1.0 / self._gradient_predivide_factor)
+
+            if self._compression != Compression.none:
+                for i, param in enumerate(self._params):
+                    if param.grad_req != 'null':
+                        param.list_grad()[0][:] = self._compression.decompress(grads[i], ctxs[i])
         else:
             # In MXNet 2.0, param.name is no longer unique.
             # Meanwhile, since horovod requires Python 3.6, there is no need to sort
             # self._params as enumerating a python dict is always deterministic.
             for i, param in enumerate(self._params):
                 if param.grad_req != 'null':
-                    allreduce_(param.list_grad()[0], average=False,
+                    tensor_compressed, ctx = self._compression.compress(param.list_grad()[0])
+                    allreduce_(tensor_compressed, average=False,
                                name=self._prefix + str(i), priority=-i,
                                prescale_factor=1.0 / self._gradient_predivide_factor)
+                    param.list_grad()[0][:] = self._compression.decompress(tensor_compressed, ctx)
```

Comment:
In the case without compression enabled (i.e. the tensor is the same on the left and right hand sides of this assignment), does MXNet do any copies or does it automatically skip the assignment? I think it may just skip the assignment so this shouldn't result in extra copies in the case without compression. With that said, it still might be good to add the if statement here to only do the copy assignment conditionally on `self._compression != Compression.none` for safety and consistency.

Reply:
Agree. I will add if statement.